File: Niveau 4/Niveau 4.0/baguenaudier.hs
--------------------------------------------------------------------------------
-- Name:        Baguenaudier
-- Purpose:     France-IOI niveau 4.0
--
-- Author:      Sébastien
--
-- Created:     19/11/2022
--------------------------------------------------------------------------------

main :: IO ()
main = do
    n <- readLn :: IO Int
    vider n
   
-- Vide les k premières cases   
vider :: Int -> IO ()
vider k
    | k > 0 = do
        vider (k-2)
        print k -- Enlève la pierre de la case n° k
        remplir (k-2)
        vider (k-1)
    | otherwise = return ()

-- Remplit les k premières cases
remplir :: Int -> IO ()
remplir k
    | k > 0 = do
        remplir (k-1)
        vider (k-2)
        print k -- Pose la pierre de la case n° k
        remplir (k-2)
    | otherwise = return ()

File: Niveau 4/Niveau 4.0/baguenaudier.py
#-------------------------------------------------------------------------------
# Name:        Baguenaudier
# Purpose:     France-IOI niveau 4.0
#
# Author:      Sébastien
#
# Created:     19/11/2022
#-------------------------------------------------------------------------------

import time

def modifier(cases, k, valeur):
    if k > 1:
        remplir(cases, k-1)
    if k > 2:
        for i in range(k-2, 0, -1):
            vider(cases, i)
    cases[k] = valeur
    print(k)

def vider(cases, k):
    if cases[k]: modifier(cases, k, False)

def remplir(cases, k):
    if not cases[k]: modifier(cases, k, True)

def main():
    n = int(input())
    cases = [True]*(n+1)
    for k in range(n, 0, -1):
        vider(cases, k)

if __name__ == '__main__':
    start = time.time()
    main()
    end = time.time()
##  print("Temps d'exécution :", end-start, "ms")
